src/metrics/include/metrics/operation.hpp

```diff
@@ -0,0 +1,260 @@
+ * the instance.
+ */
+template <typename ClockSource>
+class OperationContextT : private boost::noncopyable {
+private:
+    using OutcomeType = typename OperationEvent<ClockSource>::OutcomeType;
+
+public:
+    using time_point = typename ClockSource::time_point;
+
+    explicit OperationContextT(v1::OperationImpl<ClockSource>& op)
+        : _op{std::addressof(op)}, _started{ClockSource::now()} {}
+
+    OperationContextT(OperationContextT<ClockSource>&& other) noexcept
+        : _op{std::move(other._op)},
+          _started{std::move(other._started)},
+          _event{std::move(other._event)},
+          _isClosed{std::exchange(other._isClosed, true)} {}
+
+    ~OperationContextT() {
+        if (!_isClosed) {
+            BOOST_LOG_TRIVIAL(warning)
+                << "Metrics not reported because operation '" << _op->getOpName()
+                << "' being run by actor '" << _op->getActorName()
+                << "' did not close with success() or failure().";
+        }
+    }
+
+    /**
+     * Increments the counter for the number of iterations. This method only needs to be called if
+     * an actor is periodically reporting its operations. By default an `iters = 1` value is
+     * automatically reported.
+     */
+    void addIterations(count_type iters) {
+        _event.iters += iters;
+    }
+
+    /**
+     * Increments the counter for the number of documents inserted, modified, deleted, etc.
+     */
+    void addDocuments(count_type ops) {
+        _event.ops += ops;
+    }
+
+    /**
+     * Increments the size in bytes of the number of documents inserted, etc.
+     */
+    void addBytes(count_type size) {
+        _event.size += size;
+    }
+
+    /**
+     * Increments the counter for the number of write errors, transient transaction errors, etc.
+     */
+    void addErrors(count_type errors) {
+        _event.errors += errors;
+    }
+
+    /**
+     * Report the operation as having succeeded.
+     *
+     * @warning After calling success() it is illegal to call any methods on this instance.
+     */
+    void success() {
+        reportOutcome(OutcomeType::kSuccess);
+    }
+
+    /**
+     * Report the operation as having failed. An `errors > 0` value won't be reported unless
+     * addErrors() has already been called.
+     *
+     * @warning After calling failure() it is illegal to call any methods on this instance.
+     */
+    void failure() {
+        reportOutcome(OutcomeType::kFailure);
+    }
+
+    /**
+     * Don't report the operation.
+     *
+     * @warning After calling discard() it is illegal to call any methods on this instance.
+     */
+    void discard() {
+        _isClosed = true;
+    }
+
+private:
+    void reportOutcome(OutcomeType outcome) {
+        auto finished = ClockSource::now();
+        _event.duration = finished - _started;
+        _event.outcome = outcome;
+
+        if (_event.iters == 0) {
```

Comment:
maybe have this be `-1` to indicate unset - that way it's possible to indicate that no actual operations happened in this iteration. could be useful in cases where outcome=failure to indicate that maybe something happened during setup that prevented the operation from happening.

Reply:
Maybe @tychoish can say whether reporting `n=1` is (1) something Cedar currently supports, and (2) if so, something Cedar would intentionally keep supporting. Either way, I'm inclined to leave this as-is.

> could be useful in cases where outcome=failure to indicate that maybe something happened during setup that prevented the operation from happening.

I think I'd want to know more about the circumstances to understand why if the operation failed to happen that we'd want to report anything rather than log something and try to cause the Evergreen task to fail.

Reply:
I believe that cedar would appreciate itter counts